src/or/onion.c

```diff
@@ -1339,3 +1342,466 @@ extended_cell_format(uint8_t *command_out, uint16_t *len_out,
+    /* TODO: When we have handshake types which use wide cells, we should check
+     * that their lengths match the expected lengths here. --isis */
+  default:
+    /* If we've been given handshake_data, make sure we were given a
+     * length in (0,10240]. */
+    tor_assert(handshake_len > 0);
+    tor_assert(handshake_len <= 10240);
+
+    /* If we've been given both handshake_data and padding_data, make
+     * sure their combined reported lengths are not greater than the
+     * maximum. */
+    tor_assert(handshake_len + padding_len <= 10240);
+  }
+}
+
+/**
+ * Initialise a CREATE2V cell.
+ *
+ * Copies <b>handshake_len</b> bytes of <b>handshake_data</b> into the
+ * <b>cell_out</b>'s body.hdata member, and then copies <b>padding_len</b>
+ * bytes of <b>padding_data</b> into the <b>cell_out</b>'s body.ignored member.
+ * Also sets cell_out->body.htype to <b>handshake_type</b>.
+ */
+void
+create2v_cell_init(create2v_cell_t *cell_out,
+                   const uint16_t handshake_type,
+                   const uint8_t *handshake_data,
+                   const uint16_t handshake_len,
+                   const uint8_t *padding_data,
+                   const uint16_t padding_len)
+{
+  tor_assert(cell_out);
+
+  check_handshake_and_padding_len(handshake_type, handshake_len,
+                                  padding_len, true);
+
+  create2v_cell_body_set_htype(cell_out->body, handshake_type);
+  create2v_cell_body_set_hlen(cell_out->body, handshake_len);
+  create2v_cell_body_setlen_hdata(cell_out->body, (size_t)handshake_len);
+  create2v_cell_body_setlen_ignored(cell_out->body, (size_t)padding_len);
+
+  if (handshake_len)
+    memcpy(create2v_cell_body_getarray_hdata(cell_out->body),
+           handshake_data, handshake_len);
+  if (padding_len)
+    memcpy(create2v_cell_body_getarray_ignored(cell_out->body),
+           padding_data, padding_len);
+}
+
+/**
+ * Deallocate a CREATE2V cell.
+ */
+void
+create2v_cell_free(create2v_cell_t *cell)
+{
+  create2v_cell_body_free(cell->body);
+  tor_free(cell);
+}
+
+/**
+ * Given the Create2VMaximumData consensus parameter, calculate the
+ * total maximum allowable size of an incoming CREATE(D)2V buffer.
+ */
+static int
+create2v_cell_calculate_maximum_total(void)
+{
+  return networkstatus_get_create2v_maximum_data(NULL)
+    // XXXisis what about these extra bits of stuff? are we only counting
+    // "ignored" and "hdata" as data? or should we also count all of this?
+    + 2 // for the uint16_t htype
+    + 2 // for the uint16_t hlen
+    + sizeof(size_t) * 4 // for n_ and allocated_ in TRUNNEL_DYNARRAY_HEADs
+    + 1; // for the uint8_t trunnel_error_code_
+}
+
+/**
+ * Check that the encoded length of an incoming CREATE2V or CREATED2V cell is
+ * less than the maximum allowed.
+ *
+ * Returns true if it's okay, and false otherwise.
+ */
+static bool
+create2v_cell_check_encoded_length(const size_t len)
+{
+  const uint16_t max_len = create2v_cell_calculate_maximum_total();
```

Comment:
We should use an int or a uint or a size_t here, since the function can techincally return a value larger than uint16_max